runtime_services/src/runtime_services.rs

```diff
@@ -0,0 +1,864 @@
+#![cfg_attr(all(not(test), not(feature = "mockall")), no_std)]
+
+extern crate alloc;
+
+pub mod variable_services;
+
+#[cfg(any(test, feature = "mockall"))]
+use mockall::automock;
+
+use alloc::vec::Vec;
+use core::{
+    ffi::c_void,
+    marker::PhantomData,
+    ptr,
+    sync::atomic::{AtomicPtr, Ordering},
+};
+
+use r_efi::efi;
+pub use variable_services::{GetVariableStatus, VariableIdentifier, VariableInfo, VariableNameIterator};
+
+/// This is the runtime services used in the UEFI.
+/// it wraps an atomic ptr to [`efi::RuntimeServices`]
+#[derive(Debug)]
+pub struct StandardRuntimeServices<'a> {
+    efi_runtime_services: AtomicPtr<efi::RuntimeServices>,
+    _lifetime_marker: PhantomData<&'a efi::RuntimeServices>,
+}
+
+impl<'a> StandardRuntimeServices<'a> {
+    /// Create a new StandardRuntimeServices with the provided [efi::RuntimeServices].
+    pub const fn new(efi_runtime_services: &'a efi::RuntimeServices) -> Self {
+        // The efi::RuntimeServices is only read, that is why we use a non mutable reference.
+        Self {
+            efi_runtime_services: AtomicPtr::new(efi_runtime_services as *const _ as *mut _),
+            _lifetime_marker: PhantomData,
+        }
+    }
+
+    /// Create a new StandardRuntimeServices that is uninitialized.
+    /// The struct need to be initialize later with [Self::initialize], otherwise, subsequent call will panic.
+    pub const fn new_uninit() -> Self {
+        Self { efi_runtime_services: AtomicPtr::new(ptr::null_mut()), _lifetime_marker: PhantomData }
+    }
+
+    /// Initialize the StandardRuntimeServices with a reference to [efi::RuntimeServices].
+    /// # Panics
+    /// This function will panic if already initialize.
+    pub fn initialize(&'a self, efi_runtime_services: &'a efi::RuntimeServices) {
+        if self.efi_runtime_services.load(Ordering::Relaxed).is_null() {
+            // The efi::RuntimeServices is only read, that is why we use a non mutable reference.
+            self.efi_runtime_services.store(efi_runtime_services as *const _ as *mut _, Ordering::SeqCst)

```

Comment:
same here, with `addr_of_mut!` if it works. Last I remember, I had some issues with using it with AtomicPtrs

Reply:
Same here